File: Backend/core/events/webhooks.js
/**
 * JavaScript wrapper for BHIV Webhooks Python API
 * This module communicates with the Python FastAPI service running on port 8005
 */

const axios = require('axios');

// Configuration
const WEBHOOKS_API_URL = process.env.WEBHOOKS_API_URL || 'http://localhost:8005';

// In-memory storage for monitoring events (fallback when Python service is not running)
const monitoringEvents = [];

/**
 * Handle escalation result callback
 * @param {Object} payload - Escalation result payload
 * @returns {Promise<Object>} Response
 */
async function handle_escalation_result(payload) {
  try {
    // Try to call the Python service
    const response = await axios.post(`${WEBHOOKS_API_URL}/callbacks/escalation-result`, payload, {
      timeout: 5000
    });
    return response.data;
  } catch (error) {
    console.warn('Python Webhooks API not available, using fallback:', error.message);
    
    // Fallback implementation
    console.log('Escalation result received (fallback):', payload);
    return {
      status: 'acknowledged',
      message: 'Escalation result received (fallback mode)',
      timestamp: new Date().toISOString()
    };
  }
}

/**
 * Handle generic callback
 * @param {string} callback_type - Type of callback
 * @param {Object} payload - Callback payload
 * @returns {Promise<Object>} Response
 */
async function handle_generic_callback(callback_type, payload) {
  try {
    // Try to call the Python service
    const response = await axios.post(`${WEBHOOKS_API_URL}/callbacks/${callback_type}`, payload, {
      timeout: 5000
    });
    return response.data;
  } catch (error) {
    console.warn('Python Webhooks API not available, using fallback:', error.message);
    
    // Fallback implementation
    console.log(`Generic callback received (fallback) - Type: ${callback_type}`, payload);
    return {
      status: 'acknowledged',
      callback_type,
      message: 'Callback received (fallback mode)',
      timestamp: new Date().toISOString()
    };
  }
}

/**
 * Get monitoring events
 * @param {string} event_type - Optional event type filter
 * @returns {Promise<Object>} Monitoring events
 */
async function get_monitoring_events(event_type) {
  try {
    // Try to call the Python service
    const url = event_type 
      ? `${WEBHOOKS_API_URL}/monitoring/events?event_type=${event_type}`
      : `${WEBHOOKS_API_URL}/monitoring/events`;
    
    const response = await axios.get(url, {
      timeout: 5000
    });
    return response.data;
  } catch (error) {
    console.warn('Python Webhooks API not available, using fallback:', error.message);
    
    // Fallback implementation
    const filteredEvents = event_type
      ? monitoringEvents.filter(e => e.event_type === event_type)
      : monitoringEvents;
    
    return {
      events: filteredEvents,
      count: filteredEvents.length,
      source: 'fallback'
    };
  }
}

/**
 * Log a monitoring event
 * @param {Object} payload - Event payload
 * @returns {Promise<Object>} Response
 */
async function log_monitoring_event(payload) {
  try {
    // Try to call the Python service
    const response = await axios.post(`${WEBHOOKS_API_URL}/monitoring/events`, payload, {
      timeout: 5000
    });
    return response.data;
  } catch (error) {
    console.warn('Python Webhooks API not available, using fallback:', error.message);
    
    // Fallback implementation
    const event = {
      ...payload,
      event_id: `fallback-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`,
      timestamp: new Date().toISOString()
    };
    
    monitoringEvents.push(event);
    
    // Keep only last 1000 events
    if (monitoringEvents.length > 1000) {
      monitoringEvents.shift();
    }
    
    return {
      status: 'logged',
      event_id: event.event_id,
      message: 'Event logged (fallback mode)'
    };
  }
}

/**
 * Replay a failed event
 * @param {string} event_id - Event ID to replay
 * @returns {Promise<Object>} Response
 */
async function replay_failed_event(event_id) {
  try {
    // Try to call the Python service
    const response = await axios.post(`${WEBHOOKS_API_URL}/monitoring/replay/${event_id}`, {}, {
      timeout: 5000
    });
    return response.data;
  } catch (error) {
    if (error.response && error.response.status === 404) {
      throw new Error('Event not found');
    }
    
    console.warn('Python Webhooks API not available, using fallback:', error.message);
    
    // Fallback implementation
    const event = monitoringEvents.find(e => e.event_id === event_id);
    
    if (!event) {
      throw new Error('Event not found');
    }
    
    return {
      status: 'replayed',
      event_id,
      message: 'Event replayed (fallback mode)',
      timestamp: new Date().toISOString()
    };
  }
}

/**
 * Health check for the Webhooks service
 * @returns {Promise<Object>} Health status
 */
async function health_check() {
  try {
    // Try to call the Python service
    const response = await axios.get(`${WEBHOOKS_API_URL}/health`, {
      timeout: 5000
    });
    return response.data;
  } catch (error) {
    console.warn('Python Webhooks API not available:', error.message);
    
    // Return fallback status
    return {
      status: 'healthy',
      service: 'BHIV Webhooks API (Fallback)',
      version: '1.0.0',
      monitoring_events_count: monitoringEvents.length,
      python_service: 'unavailable'
    };
  }
}

module.exports = {
  handle_escalation_result,
  handle_generic_callback,
  get_monitoring_events,
  log_monitoring_event,
  replay_failed_event,
  health_check
};

File: Backend/core/events/webhooks.py
"""
Webhook Handler for BHIV Core System

This module handles webhook callbacks for orchestrated outcomes
and provides monitoring endpoints for failed event deliveries.
"""

from fastapi import FastAPI, HTTPException, status
from pydantic import BaseModel
from typing import List, Optional, Dict, Any
import uuid
from datetime import datetime
import logging
import json

# Set up logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

app = FastAPI(
    title="BHIV Core Webhooks",
    description="Webhook endpoints for orchestrated outcomes",
    version="1.0.0"
)

class WebhookPayload(BaseModel):
    """Payload for webhook callbacks"""
    outcomeId: str
    caseId: str
    eventType: str
    result: Dict[str, Any]
    timestamp: str

class WebhookResponse(BaseModel):
    """Response model for webhook acceptance"""
    status: str = "received"
    messageId: str

class MonitoringEvent(BaseModel):
    """Model for monitoring events"""
    eventId: str
    eventType: str
    status: str
    timestamp: str
    details: Optional[str] = None

# In-memory storage for webhook events and monitoring (in production, this would be a database)
webhook_events = []
monitoring_events = []

@app.post("/callbacks/escalation-result", response_model=WebhookResponse)
async def handle_escalation_result(payload: WebhookPayload):
    """
    Handle escalation result callbacks from orchestration.
    """
    try:
        # Generate message ID
        message_id = str(uuid.uuid4())
        
        # Store the webhook event
        event_data = {
            "messageId": message_id,
            "payload": payload.dict(),
            "receivedAt": datetime.now().isoformat()
        }
        webhook_events.append(event_data)
        
        logger.info(f"Received escalation result webhook: {payload}")
        
        # In a real implementation, process the escalation result here
        # For example, update case status, notify stakeholders, etc.
        
        return WebhookResponse(messageId=message_id)
    except Exception as e:
        logger.error(f"Error handling escalation result: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to process webhook: {str(e)}"
        )

@app.post("/callbacks/{callback_type}", response_model=WebhookResponse)
async def handle_generic_callback(callback_type: str, payload: WebhookPayload):
    """
    Handle generic callback events.
    """
    try:
        # Generate message ID
        message_id = str(uuid.uuid4())
        
        # Store the webhook event
        event_data = {
            "messageId": message_id,
            "callbackType": callback_type,
            "payload": payload.dict(),
            "receivedAt": datetime.now().isoformat()
        }
        webhook_events.append(event_data)
        
        logger.info(f"Received {callback_type} webhook: {payload}")
        
        return WebhookResponse(messageId=message_id)
    except Exception as e:
        logger.error(f"Error handling {callback_type} callback: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to process webhook: {str(e)}"
        )

@app.get("/monitoring/events")
async def get_monitoring_events(event_type: Optional[str] = None):
    """
    Get monitoring events, optionally filtered by event type.
    """
    if event_type:
        filtered_events = [event for event in monitoring_events if event.get("eventType") == event_type]
        return filtered_events
    return monitoring_events

@app.post("/monitoring/events")
async def log_monitoring_event(event: MonitoringEvent):
    """
    Log a monitoring event.
    """
    monitoring_events.append(event.dict())
    logger.info(f"Logged monitoring event: {event}")
    return {"status": "logged"}

@app.post("/monitoring/replay/{event_id}")
async def replay_failed_event(event_id: str):
    """
    Replay a failed event delivery.
    """
    # Find the event in webhook events
    event_to_replay = None
    for event in webhook_events:
        if event.get("messageId") == event_id:
            event_to_replay = event
            break
    
    if not event_to_replay:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Event not found"
        )
    
    # In a real implementation, we would replay the event to the original destination
    # For now, we'll just log that a replay was requested
    logger.info(f"Replay requested for event {event_id}")
    
    # Log this as a monitoring event
    monitoring_event = {
        "eventId": str(uuid.uuid4()),
        "eventType": "replay",
        "status": "initiated",
        "timestamp": datetime.now().isoformat(),
        "details": f"Replay initiated for event {event_id}"
    }
    monitoring_events.append(monitoring_event)
    
    return {
        "status": "replay_initiated",
        "eventId": event_id,
        "monitoringEventId": monitoring_event["eventId"]
    }

@app.get("/health")
async def health_check():
    """Health check endpoint."""
    return {
        "status": "healthy",
        "service": "BHIV Core Webhooks",
        "version": "1.0.0",
        "webhook_events_count": len(webhook_events),
        "monitoring_events_count": len(monitoring_events)
    }

if __name__ == "__main__":
    import uvicorn
    import argparse
    
    parser = argparse.ArgumentParser(description="BHIV Core Webhooks")
    parser.add_argument("--port", type=int, default=8005, help="Port to run the server on (default: 8005)")
    parser.add_argument("--host", type=str, default="0.0.0.0", help="Host to run the server on (default: 0.0.0.0)")
    args = parser.parse_args()
    
    print("\n" + "="*60)
    print("  BHIV CORE WEBHOOKS")
    print("="*60)
    print(f" Server URL: http://{args.host}:{args.port}")
    print(f" API Documentation: http://{args.host}:{args.port}/docs")
    print("\n Endpoints:")
    print("   POST /callbacks/escalation-result - Handle escalation results")
    print("   POST /callbacks/{callback_type} - Handle generic callbacks")
    print("   GET /monitoring/events - Get monitoring events")
    print("   POST /monitoring/events - Log monitoring events")
    print("   POST /monitoring/replay/{event_id} - Replay failed events")
    print("   GET /health - Health check")
    print("="*60)
    
    uvicorn.run(app, host=args.host, port=args.port)